File: android-samples/poslink/EposPosLinkExample/app/src/main/java/com/example/eposappexample/poslink/ui/MainViewModel.kt
package com.example.eposappexample.poslink.ui

import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.example.eposappexample.poslink.TeyaUtils
import com.example.eposappexample.poslink.models.Product

class MainViewModel : ViewModel() {

    var basket by mutableStateOf(listOf<Product>())
        private set

    var tipInput by mutableStateOf("")
        private set

    val tipAmount by derivedStateOf { tipInput.toDoubleOrNull() ?: 0.0 }

    val subtotal by derivedStateOf { basket.sumOf { it.price * it.quantity } }

    val total by derivedStateOf { subtotal + tipAmount }

    val itemCount by derivedStateOf { basket.sumOf { it.quantity } }

    val payEnabled by derivedStateOf { basket.isNotEmpty() }

    fun addProduct(product: Product) {
        val existing = basket.find { it.id == product.id }
        basket = if (existing != null) {
            basket.map { if (it.id == product.id) it.copy(quantity = it.quantity + 1) else it }
        } else {
            basket + product.copy(quantity = 1)
        }
    }

    fun removeProduct(product: Product) {
        val existing = basket.find { it.id == product.id } ?: return
        basket = if (existing.quantity > 1) {
            basket.map { if (it.id == product.id) it.copy(quantity = it.quantity - 1) else it }
        } else {
            basket.filter { it.id != product.id }
        }
    }

    fun updateTipInput(value: String) {
        tipInput = value
    }

    fun pay() {
        val totalInMinorUnits = (total * 100).toInt()
        val tipInMinorUnits = (tipAmount * 100).toInt()
        TeyaUtils.makePayment(totalInMinorUnits, tipInMinorUnits)
    }

    fun printReceipt() {
        TeyaUtils.printReceipt(basket, tipAmount)
    }

    fun clearUserAuth() {
        TeyaUtils.clearUserAuth()
        TeyaUtils.setUp()
    }

    fun clearDeviceLink() {
        TeyaUtils.clearDeviceLink()
        TeyaUtils.setUp()
    }
}

File: ios-samples/poslink/EposPosLinkExample/EposPosLinkExample/ViewModels/MainViewModel.swift
import Foundation

@Observable
final class MainViewModel {
    var basket: [Product] = []
    var tipInput: String = ""

    var tipAmount: Double {
        Double(tipInput.replacingOccurrences(of: ",", with: ".")) ?? 0.0
    }

    var subtotal: Double {
        basket.reduce(0.0) { $0 + $1.price * Double($1.quantity) }
    }

    var total: Double {
        subtotal + tipAmount
    }

    var itemCount: Int {
        basket.reduce(0) { $0 + $1.quantity }
    }

    var payEnabled: Bool {
        !basket.isEmpty
    }

    func addProduct(_ product: Product) {
        if let index = basket.firstIndex(where: { $0.id == product.id }) {
            basket[index].quantity += 1
        } else {
            var newProduct = product
            newProduct.quantity = 1
            basket.append(newProduct)
        }
    }

    func removeProduct(_ product: Product) {
        guard let index = basket.firstIndex(where: { $0.id == product.id }) else { return }
        if basket[index].quantity > 1 {
            basket[index].quantity -= 1
        } else {
            basket.remove(at: index)
        }
    }

    func updateTipInput(_ value: String) {
        tipInput = value
    }

    func pay() {
        let totalInMinorUnits = Int32((total * 100).rounded())
        let tipInMinorUnits = Int32((tipAmount * 100).rounded())
        TeyaService.shared.makePayment(totalMinorUnits: totalInMinorUnits, tipMinorUnits: tipInMinorUnits)
    }

    func printReceipt() {
        TeyaService.shared.printReceipt(products: basket, tip: tipAmount)
    }

    func clearUserAuth() {
        TeyaService.shared.clearUserAuth()
        TeyaService.shared.setUp()
    }

    func clearDeviceLink() {
        TeyaService.shared.clearDeviceLink()
        TeyaService.shared.setUp()
    }
}
